### DISCUSSION TITLE
onMobDeath details parameters

### DISCUSSION THREAD

**Discussion by zach2good** [topaz-next/topaz#2599] [Architecture]
As mentioned in this PR: <https://github.com/topaz-next/topaz/pull/2569>

We're starting to convert KO's work on Assaults, and Eden's `onMobDeath` has an additional argument:
`entity.onMobDeath = function(mob, player, isKiller, firstCall)`

Here, we have `isKiller` and `noKiller`

This started the conversation, how do we fix/extend this so it's clean and covers all bases?

I think we're heading towards a params/details/args type of object, with all of the details populated.

```lua
-- Called once PER CHAR IN THE KILLING ALLIANCE
-- Use params to filter out the logic you want to apply, when you want to apply it
onMobDeath(mob, player, params)
...
params = {
    isKiller -- True for the char with the killing blow
    noKiller -- True if there was no killing blow (DoT death, GM commands etc.)
    firstCall -- True when being called for the first time (potentially of many) 
    isWSKill -- True when the mob was killed by a Weaponskill
    isMagicKill -- True when the mob was killed by a direct damage spell
    etc.
}
```

**Comment by jefferyrlc** [comment]
Should probably have a parameter for mob deaths that have a proper proc before kill, such as abyssea NMs.

**Reply by TeoTwawki** [reply]
not sure that'll be needed but woudl be trivially added. I'm thinking that would likely be tracked on its own because stuff like loot pool changes simply happening on proc, before its dead

**Comment by TeoTwawki** [comment]
"firstCall" : lets just have a count rather than noting its first, so it can be used for more. Right now we can establish a run once easily without it, but I could do a few things with a count against how many times it ran. Say I need to randomly select a party member on mob death - I can use a random number and do something on the nTh call based on that.

**Reply by cocosolos** [reply]
You could do something like this now on the first call with something like
```lua
if isKiller then
    local party = player:getParty()
    party[math.random(#party)]:addItem(itemid)
end
```
